### PR TITLE
Add http/https ProxyUrl setting in ConnectionPolicy

### PR DESCRIPTION
**Pull request by ziyel**
I'd like to add http/https proxy support for NodeJS SDK. 

The codes of PR is edited from PR https://github.com/Azure/azure-documentdb-node/pull/124.
As the PR https://github.com/Azure/azure-documentdb-node/pull/124 has not been merged for months, I raised this PR. 

@ryanmentzer , if you have time to continue your PR https://github.com/Azure/azure-documentdb-node/pull/124, please let me know, then I will close this PR. Otherwise I will try to merge this.

### CODE REVIEW

**Comment by ryanmentzer** [comment]
@ziyel please merge yours. Mine stalled out while getting submission approval from a previous employer - the view didn't end up being worth the climb. Thanks for carrying it forward!

**Comment by ziyel** [comment]
Thanks @ryanmentzer

**Comment by ziyel** [comment]
@rnagpal @moderakh Can someone help to review the PR?

**Comment by moderakh** [comment]
Hey @ziyel I tried to run the tests internally and the new test you added fails:

```
Error: done() invoked with non-Error: Should create database in error while the proxy setting is not correct
at test\proxyTests.js:78:21
at lib\documentclient.js:2175:21
at lib\retryUtility.js:98:20
at IncomingMessage.<anonymous> (lib\request.js:128:13)
at endReadableNT (_stream_readable.js:974:12)
at _combinedTickCallback (internal/process/next_tick.js:80:11)
at process._tickDomainCallback (internal/process/next_tick.js:128:9)
```

Can you please verify your test?

**Comment by ziyel** [comment]
@moderakh The test can pass when run independently.
```
$ mocha -t 0 proxyTests.js

  Validate http proxy setting in environment variable
    √ nativeApi Client Should successfully execute request (104ms)
    √ nativeApi Client Should execute request in error while the proxy setting is not correct (2008ms)


  2 passing (2s)
```
The proxy environment need to be set when initialize the `request` module.  The test will fail, because all tests are run in the same process.

**Comment by ziyel** [comment]
For now, the request agent is a global instance https://github.com/Azure/azure-documentdb-node/blob/master/source/lib/request.js#L33.

All `DocumentClient` will use the some agent to send the request which cause the proxy test failed when run together with other tests. 
And I think it's not make sense to share the same agent within all clients.
I'd like to raise a seperate PR to refactor this and we can also make the proxy setting plug-gable after refactoring. 
Could we merge this PR first? As the test cases of proxy are actually passed.

**Comment by ziyel** [comment]
@moderakh , I have make the code change. Could you please help to review?

File: source/lib/documentclient.js [hunk 1]
```diff
@@ -2257,27 +2280,27 @@ var DocumentClient = Base.defineClass(
 
         /** @ignore */
         get: function (url, request, headers, callback) {
-            return RequestHandler.request(this._globalEndpointManager, this.connectionPolicy, "GET", url, request, undefined, this.defaultUrlParams, headers, callback);
+            return RequestHandler.request(this._globalEndpointManager, this.connectionPolicy, this.requestAgent, "GET", url, request, undefined, this.defaultUrlParams, headers, callback);
```
Comment (moderakh) [inline source/lib/documentclient.js]: connectionPolicy (which has proxy config) is passed to  RequestHandler.request, is it possible to move all the logic for proxy creation inside RequestHandler.request ?

You can create requestAgent inside RequestHandler.request too.l

File: source/lib/documentclient.js [hunk 2]
```diff
@@ -97,6 +100,26 @@ var DocumentClient = Base.defineClass(
...
+            if (!!proxyUrl.auth) {
+                requestAgentOptions.proxy.proxyAuth = proxyUrl.auth;
+            }
+
+            this.requestAgent = proxyUrl.protocol.toLowerCase() === "https:" ?
+                tunnel.httpsOverHttps(requestAgentOptions) :
+                tunnel.httpsOverHttp(requestAgentOptions);
+        } else {
+            this.requestAgent = new https.Agent(requestAgentOptions);
+        }
```
Comment (moderakh) [inline source/lib/documentclient.js]: connectionPolicy (which has proxy config) is passed to  RequestHandler.request, is it possible to move all the logic for proxy creation inside RequestHandler.request ?

You can create requestAgent inside RequestHandler.request.

Comment (ziyel) [reply source/lib/documentclient.js]: @moderakh 
If we create the request agent inside the `RequestHandler.request`, we will send every request with a new request agent. In this case, the `KeepAlive` option will not work.
As the API is using `https`, if we are not using `KeepAlive` agent, the performance will decrease significantly. Thus I suggest we create the request agent in the document client.

Comment (ziyel) [reply source/lib/documentclient.js]: Another choice is making the `requestAgent` as a singleton in `RequestHandler.request`. But in this case, all the document client will share the same `requestAgent` which I think is not make sense.

Comment (ziyel) [reply source/lib/documentclient.js]: @moderakh 
Is there any actions I need to perform to merge the PR?

**Comment by ziyel** [comment]
@moderakh
Is there any actions I need to perform to merge the PR?

**Comment by moderakh** [comment]
Thanks @ziyel for your contribution. I merged your PR.

**Pull request merged**